builder/pipeline.cc

```diff
@@ -576,21 +695,75 @@ class pipeline_builder {
       result = add_input_crops(result, base_f);
     }
 
+    symbol_map<symbol_id> uncropped_subs;
     // Add all allocations at this loop level.
     for (int ix = order_.size() - 1; ix >= 0; ix--) {
       const auto& f = order_[ix];
       for (const auto& o : f->outputs()) {
         const auto& b = o.buffer;
-        if (allocated.count(b)) continue;
+        if (output_syms_.count(b->sym())) continue;
 
         if ((b->store_at() && *b->store_at() == at) || (!b->store_at() && at.root())) {
-          result = allocate::make(b->sym(), b->storage(), b->elem_size(), b->dims(), result);
+          symbol_id uncropped = ctx.insert_unique(ctx.name(b->sym()) + ".uncropped");
+          uncropped_subs[b->sym()] = uncropped;
+          result = clone_buffer::make(uncropped, b->sym(), result);
+
+          const auto& dims = *inferred_dims_[b->sym()];
+          const auto& bounds = *allocation_bounds_[b->sym()];
+          result = allocate::make(b->sym(), b->storage(), b->elem_size(), dims, result);
+
+          std::vector<stmt> checks;
+          for (std::size_t d = 0; d < dims.size(); ++d) {
+            if (d < bounds.size()) {
+              checks.push_back(check::make(dims[d].min() <= bounds[d].min));
+              checks.push_back(check::make(dims[d].max() >= bounds[d].max));
+            }
+          }
+
+          result = block::make(std::move(checks), result);
         }
       }
     }
 
+    // Substitute references to the intermediate buffers with the 'name.uncropped' when they
+    // are used as an input arguments.
+    result = substitute_uncropped(result, uncropped_subs);
```

Comment:
I think you should do this substitution one at a time as needed, so it only substitutes exactly within the body of the relevant `allocate`. I don't think we need to worry about shadowing here yet, but I'd rather not have to think about it. We can do this as an optimization later if needed.

Reply:
This does it kind of locally already: the code here creates allocates which are stored at this specific loop level and then substitutes only within them.

Reply:
Not sure I follow, doesn't the loop above make a bunch of nested allocates, and then substitute all the uncropped buffers afterwards? I basically want to only substitute an uncropped buffer into the scope where the clone_buffer exists, otherwise we risk introducing a use of the cloned buffer before it exists, leading to either an undefined symbol crash or a more subtle bug.

That said: I guess this can't happen unless the same symbol_id is written by two different funcs, which isn't allowed? Maybe comment to this affect why it's safe to substitute all these symbols at once?